File: functions/src/__tests__/soap_section_catalog.test.ts
import {
  SOAP_LAST_REVIEWED,
  SOAP_SCHEMA_VERSION,
  SOAP_SECTION_TEMPERATURE,
  SOAP_SECTIONS,
  jsonSchemaForSection,
  soapSectionByName,
} from "../lib/soap_section_catalog";

describe("SOAP catalog invariants", () => {
  test("contains the four canonical sections, in order", () => {
    expect(SOAP_SECTIONS.map((s) => s.section)).toEqual([
      "subjective",
      "objective",
      "assessment",
      "plan",
    ]);
  });

  test("every section has at least one required field", () => {
    for (const spec of SOAP_SECTIONS) {
      const reqs = spec.fields.filter((f) => f.required);
      expect(reqs.length).toBeGreaterThan(0);
    }
  });

  test("every section has a non-empty regulatoryRefs anchor", () => {
    for (const spec of SOAP_SECTIONS) {
      expect(spec.regulatoryRefs.length).toBeGreaterThan(0);
    }
  });

  test("every section has temperature in [0,1]", () => {
    for (const spec of SOAP_SECTIONS) {
      const t = SOAP_SECTION_TEMPERATURE[spec.section];
      expect(t).toBeGreaterThanOrEqual(0);
      expect(t).toBeLessThanOrEqual(1);
    }
  });

  test("field keys are unique within each section", () => {
    for (const spec of SOAP_SECTIONS) {
      const seen = new Set<string>();
      for (const f of spec.fields) {
        expect(seen.has(f.key)).toBe(false);
        seen.add(f.key);
      }
    }
  });

  test("schema version pinned to a positive integer", () => {
    expect(Number.isInteger(SOAP_SCHEMA_VERSION)).toBe(true);
    expect(SOAP_SCHEMA_VERSION).toBeGreaterThan(0);
  });

  test("lastReviewed is YYYY-MM", () => {
    expect(SOAP_LAST_REVIEWED).toMatch(/^\d{4}-\d{2}$/);
  });

  test("assessment.risk_assessment is required + citation-required", () => {
    const spec = soapSectionByName("assessment");
    const risk = spec.fields.find((f) => f.key === "risk_assessment");
    expect(risk).toBeDefined();
    expect(risk?.required).toBe(true);
    expect(risk?.citationRequired).toBe(true);
  });

  test("plan carries safety_plan_reference", () => {
    const spec = soapSectionByName("plan");
    expect(spec.fields.find((f) => f.key === "safety_plan_reference"))
      .toBeDefined();
  });
});

describe("jsonSchemaForSection", () => {
  test("emits required-only keys for the section", () => {
    const spec = soapSectionByName("subjective");
    const schema = jsonSchemaForSection(spec) as {
      required: string[];
      properties: Record<string, unknown>;
    };
    expect(schema.required.sort()).toEqual([
      "chief_complaint",
      "history_present_illness",
      "patient_reported_symptoms",
    ]);
    expect(Object.keys(schema.properties).length).toBe(spec.fields.length);
  });

  test("citation-required fields require transcript_spans", () => {
    const spec = soapSectionByName("subjective");
    const schema = jsonSchemaForSection(spec) as {
      properties: Record<string, { required: string[] }>;
    };
    const cc = schema.properties["chief_complaint"];
    expect(cc.required).toEqual(["value", "transcript_spans"]);
  });

  test("non-citation fields only require value", () => {
    const spec = soapSectionByName("objective");
    const schema = jsonSchemaForSection(spec) as {
      properties: Record<string, { required: string[] }>;
    };
    const oms = schema.properties["outcome_measure_scores"];
    expect(oms.required).toEqual(["value"]);
  });
});

describe("soapSectionByName", () => {
  test("throws on unknown", () => {
    expect(() =>
      soapSectionByName("nonsense" as unknown as "plan"),
    ).toThrow();
  });
});

File: test/soap_section_catalog_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:psyclinicai/services/ai_scribe/soap_section_catalog.dart';

void main() {
  group('SoapSectionCatalog invariants', () {
    test('all four SOAP sections present, in canonical order', () {
      final order = SoapSectionCatalog.sections.map((s) => s.section).toList();
      expect(order, [
        SoapSection.subjective,
        SoapSection.objective,
        SoapSection.assessment,
        SoapSection.plan,
      ]);
    });

    test('every section has at least one required field', () {
      for (final spec in SoapSectionCatalog.sections) {
        final reqs = spec.fields.where((f) => f.required).toList();
        expect(
          reqs.isNotEmpty,
          isTrue,
          reason: '${spec.title} has no required field',
        );
      }
    });

    test('every section has a non-empty regulatory anchor', () {
      for (final spec in SoapSectionCatalog.sections) {
        expect(
          spec.regulatoryRefs.isNotEmpty,
          isTrue,
          reason: '${spec.title} missing regulatory anchor',
        );
      }
    });

    test('every section has temperature in [0.0, 1.0]', () {
      for (final spec in SoapSectionCatalog.sections) {
        final t = SoapSectionCatalog.sectionTemperature[spec.section];
        expect(t, isNotNull, reason: '${spec.title} missing temperature');
        expect(t! >= 0.0 && t <= 1.0, isTrue);
      }
    });

    test('field keys are unique within each section', () {
      for (final spec in SoapSectionCatalog.sections) {
        final keys = spec.fields.map((f) => f.key).toSet();
        expect(
          keys.length,
          spec.fields.length,
          reason: '${spec.title} has duplicate field keys',
        );
      }
    });

    test('Assessment carries risk_assessment as required', () {
      final spec = SoapSectionCatalog.bySection(SoapSection.assessment);
      final risk = spec.fields.firstWhere((f) => f.key == 'risk_assessment');
      expect(risk.required, isTrue);
      expect(risk.citationRequired, isTrue);
    });

    test('Plan carries safety_plan_reference (req-on-elevated)', () {
      final spec = SoapSectionCatalog.bySection(SoapSection.plan);
      final keys = spec.fields.map((f) => f.key).toList();
      expect(keys.contains('safety_plan_reference'), isTrue);
    });

    test('schema version is positive int', () {
      expect(SoapSectionCatalog.schemaVersion > 0, isTrue);
    });

    test('lastReviewed matches YYYY-MM', () {
      expect(
        SoapSectionCatalog.lastReviewed,
        matches(RegExp(r'^\d{4}-\d{2}$')),
      );
    });
  });

  group('isSectionComplete', () {
    test('returns false when a required field is missing', () {
      final spec = SoapSectionCatalog.bySection(SoapSection.subjective);
      expect(isSectionComplete(spec, const {}), isFalse);
    });

    test('returns false when a required string is whitespace-only', () {
      final spec = SoapSectionCatalog.bySection(SoapSection.subjective);
      expect(
        isSectionComplete(spec, const {
          'chief_complaint': '   ',
          'history_present_illness': 'x',
          'patient_reported_symptoms': ['x'],
        }),
        isFalse,
      );
    });

    test('returns true when every required field has content', () {
      final spec = SoapSectionCatalog.bySection(SoapSection.subjective);
      expect(
        isSectionComplete(spec, const {
          'chief_complaint': 'Anxiety',
          'history_present_illness': '2-week onset',
          'patient_reported_symptoms': ['Sleep loss'],
        }),
        isTrue,
      );
    });
  });

  group('soapDraftCacheKey', () {
    test('embeds schema version, tenant, section', () {
      final key = soapDraftCacheKey(
        tenantId: 'tenant-x',
        section: SoapSection.plan,
      );
      expect(key, 'soap:v1:tenant-x:plan');
    });
  });
}
